### PR TITLE
feat(ct): overhaul deployment process

### PR DESCRIPTION
**Pull request by smartcontracts**
**Description**
Overhauls the contract deployment process to use a simpler typed
deployment method. Removes the need for deployment bash scripts and
makes review of deployment configurations much easier.

TODO:
- [x] Add parsing/validation for config files
- [x] Add checks to confirm environment variables are properly set
- [x] Check if anything needs to be modified in stackman
- [x] Update README to reflect new deploy process

### CODE REVIEW

File: packages/contracts/deploy/004-OVM_CanonicalTransactionChain.deploy.ts [hunk 1]
```diff
@@ -5,10 +5,14 @@ import { DeployFunction } from 'hardhat-deploy/dist/types'
 import {
   deployAndVerifyAndThen,
   getContractFromArtifact,
+  getDeployConfig,
 } from '../src/deploy-utils'
 import { names } from '../src/address-names'
 
 const deployFn: DeployFunction = async (hre) => {
+  const deployConfig = getDeployConfig(hre.network.name)
+  console.log(deployConfig)
```
Comment (tynes) [inline packages/contracts/deploy/004-OVM_CanonicalTransactionChain.deploy.ts:14]: dead console log

File: packages/contracts/tasks/take-dump.ts [hunk 2]
```diff
@@ -1,83 +1,70 @@
...
+      gasPrice: deployConfig.gasPriceOracleL2GasPrice || 1,
+      l1BaseFee: deployConfig.gasPriceOracleL1BaseFee || 1,
+      overhead: deployConfig.gasPriceOracleOverhead || 2750,
+      scalar: deployConfig.gasPriceOracleScalar || 1_500_000,
+      decimals: deployConfig.gasPriceOracleDecimals || 6,
     },
     L2StandardBridge: {
-      l1TokenBridge: cfg.l1StandardBridgeAddress,
+      l1TokenBridge: (
+        await getContractFromArtifact(
```
Comment (tynes) [inline packages/contracts/tasks/take-dump.ts:59]: This is no longer configurable?

Comment (smartcontracts) [reply packages/contracts/tasks/take-dump.ts:59]: Correct, it gets pulled automatically from the deployment folder based on the current network name. This is effectively what we did anyway, except we did it manually which is error prone.

**Comment by tynes** [comment]
I would say at this point, the most confusing part of the deploy setup is this:
https://github.com/ethereum-optimism/optimism/blob/5c83435e77ce6a38b1207d52e6ab442c72084798/packages/contracts/hardhat.config.ts#L137-L140

These env vars modifying the networks is a bit confusing

**Comment by smartcontracts** [comment]
> I would say at this point, the most confusing part of the deploy setup is this:
> 
> https://github.com/ethereum-optimism/optimism/blob/5c83435e77ce6a38b1207d52e6ab442c72084798/packages/contracts/hardhat.config.ts#L137-L140
> 
> These env vars modifying the networks is a bit confusing

I agree. We might as well use a more standard setup (different RPC URLs for different networks? API keys? idk).

File: packages/contracts/README.md [hunk 3]
```diff
@@ -46,107 +48,105 @@ cd contracts
...
 ### Measuring test coverage:
+
 ```shell
 yarn test:coverage
 ```
 
 The output is most easily viewable by opening the html file in your browser:
+
 ```shell
 open ./coverage/index.html
```
Comment (tynes) [inline packages/contracts/README.md:79]: `open` is a mac specific thing, one particular flavor of linux uses `xdg-open`

Comment (smartcontracts) [reply packages/contracts/README.md:79]: I'll fix this in another PR

File: packages/contracts/deploy-config/mainnet.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import { DeployConfig } from '../src/deploy-config'
+
```
Comment (mslipper) [inline packages/contracts/deploy-config/mainnet.ts:2]: We'll have to add one of these for nightly, but we can do this later.

**Comment by mslipper** [comment]
Do we still have the ability to automatically set the chugsplash proxy owner? Without that capability, deployments get stuck during automated but non-CI builds like nightly.

**Review by mslipper — APPROVED** [review]
Approved conditionally on my above comment.

**Comment by smartcontracts** [comment]
> Do we still have the ability to automatically set the chugsplash proxy owner? Without that capability, deployments get stuck during automated but non-CI builds like nightly.

I haven't changed anything on that front. I think as long as you set the `AUTOMATICALLY_TRANSFER_OWNERSHIP` environment variable you should be fine.

**Comment by mslipper** [comment]
Great, then I'm happy to merge this if you ware.

File: packages/contracts/.env.example [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+CONTRACTS_TARGET_NETWORK=
+
+# Private key that will send deployment transactions
+CONTRACTS_DEPLOYER_KEY=
+
+# RPC URL connected to the L1 chain we're deploying to
+CONTRACTS_RPC_URL=
+
+# Your Etherscan API key for the L1 network
+ETHERSCAN_API_KEY=
```
Comment (maurelian) [inline packages/contracts/.env.example:11]: After a quick text search, I don't see any of these 3 values used anywhere outside the deleted bash scripts.

Can these be deleted? Where would these values now be specified? I don't see where that would be done in the deploy configs.

Comment (maurelian) [reply packages/contracts/.env.example:11]: Ah, looks like they're already in hardhat.config.ts

File: packages/contracts/scripts/deploy-scripts/mainnet.sh [hunk 6]
```diff
@@ -1,39 +0,0 @@
...
- --ctc-l2-gas-discount-divisor 32 \
- --ctc-enqueue-gas-cost 60000 \
- --scc-fraud-proof-window 604800 \
- --scc-sequencer-publish-window 12592000 \
- --ovm-sequencer-address 0x6887246668a3b87F54DeB3b94Ba47a6f63F32985 \
- --ovm-proposer-address 0x473300df21D047806A082244b417f96b32f13A33 \
- --ovm-address-manager-owner 0x9BA6e03D8B90dE867373Db8cF1A58d2F7F006b3A \
- --gasprice 150000000000 \
- --num-deploy-confirmations 4 \
- --tags upgrade \
```
Comment (maurelian) [inline packages/contracts/scripts/deploy-scripts/mainnet.sh:33]: I don't see this  `--tags` argument reflected in the `mainnet.ts` flag that replaces this. Should it be?

**Review by maurelian — COMMENTED** [review]
I see that this was merged... but I had a question about the tags flag not being include in `mainnet.ts`. 
If the point of these scripts is to make previous deployments reproducible, I think that's necessary.

**Pull request merged**